selfdrive/modeld/modeld.py

```diff
@@ -0,0 +1,271 @@
+from selfdrive.modeld.models.commonmodel_pyx import ModelFrame, CLContext, Runtime # pylint: disable=import-error, no-name-in-module
+from selfdrive.modeld.models.driving_pyx import PublishState, create_model_msg, create_pose_msg # pylint: disable=import-error, no-name-in-module
+from system.hardware import PC
+
+FEATURE_LEN = 128
+HISTORY_BUFFER_LEN = 99
+DESIRE_LEN = 8
+TRAFFIC_CONVENTION_LEN = 2
+DRIVING_STYLE_LEN = 12
+NAV_FEATURE_LEN = 256
+OUTPUT_SIZE = 5990
+MODEL_OUTPUT_SIZE = 6120
+MODEL_FREQ = 20
+
+MODEL_WIDTH = 512
+MODEL_HEIGHT = 256
+MODEL_FRAME_SIZE = MODEL_WIDTH * MODEL_HEIGHT * 3 // 2
+BUF_SIZE = MODEL_FRAME_SIZE * 2
+
+# NOTE: These are almost exactly the same as the numbers in modeld.cc, but to get perfect equivalence we might have to copy them exactly
+calib_from_medmodel = np.linalg.inv(medmodel_frame_from_calib_frame[:, :3])
+calib_from_sbigmodel = np.linalg.inv(sbigmodel_frame_from_calib_frame[:, :3])
+
+def update_calibration(device_from_calib_euler:np.ndarray, wide_camera:bool, bigmodel_frame:bool) -> np.ndarray:
+  cam_intrinsics = tici_ecam_intrinsics if wide_camera else tici_fcam_intrinsics
+  calib_from_model = calib_from_sbigmodel if bigmodel_frame else calib_from_medmodel
+  device_from_calib = rot_from_euler(device_from_calib_euler)
+  camera_from_calib = cam_intrinsics @ view_frame_from_device_frame @ device_from_calib
+  warp_matrix: np.ndarray = camera_from_calib @ calib_from_model
+  return warp_matrix
+
+class FrameMeta:
+  frame_id: int = 0
+  timestamp_sof: int = 0
+  timestamp_eof: int = 0
+
+  def __init__(self, vipc=None):
+    if vipc is not None:
+      self.frame_id, self.timestamp_sof, self.timestamp_eof = vipc.frame_id, vipc.timestamp_sof, vipc.timestamp_eof
+
+class ModelState:
+  frame: ModelFrame
+  wide_frame: ModelFrame
+  inputs: Dict[str, np.ndarray]
+  output: np.ndarray
+  prev_desire: np.ndarray  # for tracking the rising edge of the pulse
+  model: ONNXModel
+
+  def __init__(self, context:CLContext):
+    self.frame = ModelFrame(context)
+    self.wide_frame = ModelFrame(context)
+    self.prev_desire = np.zeros(DESIRE_LEN, dtype=np.float32)
+    self.output = np.zeros(MODEL_OUTPUT_SIZE, dtype=np.float32)
+    self.inputs = {
+      'desire_pulse': np.zeros(DESIRE_LEN * (HISTORY_BUFFER_LEN+1), dtype=np.float32),
+      'traffic_convention': np.zeros(TRAFFIC_CONVENTION_LEN, dtype=np.float32),
+      'nav_features': np.zeros(NAV_FEATURE_LEN, dtype=np.float32),
+      'feature_buffer': np.zeros(HISTORY_BUFFER_LEN * FEATURE_LEN, dtype=np.float32),
+    }
+
+    self.model = ONNXModel("models/supercombo.onnx", self.output, Runtime.GPU, False, context)
+    self.model.addInput("input_imgs", None)
+    self.model.addInput("big_input_imgs", None)
+    for k,v in self.inputs.items():
+      self.model.addInput(k, v)
+
+  def eval(self, buf:VisionBuf, wbuf:VisionBuf, transform:np.ndarray, transform_wide:np.ndarray, inputs:Dict[str, np.ndarray], prepare_only:bool) -> Optional[np.ndarray]:
```

Comment:
```suggestion
  def eval(self, buf: VisionBuf, wbuf: VisionBuf, transform: np.ndarray, transform_wide: np.ndarray, inputs: Dict[str, np.ndarray], prepare_only: bool) -> Optional[np.ndarray]:
```

think there's always a space here, even though equals don't have spaces